File: leetcode-cpp/medium/sortColors.cpp
#include <bits/stdc++.h>
using namespace std;

void sortColors(vector<int> &nums)
{
    if (nums.size() <= 1)
    {
        return;
    }
    int mid = nums.size() / 2;
    vector<int> leftArr(nums.begin(), nums.begin() + mid);
    vector<int> rightArr(nums.begin() + mid, nums.end());

    sortColors(leftArr);
    sortColors(rightArr);
    merger(nums, leftArr, rightArr);
}

void merger(vector<int> &nums, const vector<int> &leftArr, const vector<int> &rightArr)
{
    int i = 0, l = 0, r = 0;

    while (l < leftArr.size() && r < rightArr.size())
    {
        if (leftArr[l] < rightArr[r])
        {
            nums[i++] = leftArr[l++];
        }
        else
        {
            nums[i++] = rightArr[r++];
        }
    }
    while (l < leftArr.size())
    {
        nums[i++] = leftArr[l++];
    }
    while (r < rightArr.size())
    {
        nums[i++] = rightArr[r++];
    }
}

int main()
{

    /*
    Given an array nums with n objects colored red, white, or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white, and blue.

    We will use the integers 0, 1, and 2 to represent the color red, white, and blue, respectively.

    You must solve this problem without using the library's sort function.
    */
    vector<int> nums = {2, 0, 2, 1, 1, 0};

    sortColors(nums);

    for (int i = 0; i < nums.size(); i++)
    {
        cout << nums[i] << " ";
    }

    return 0;
}

File: leetcode-java/medium/sortColors.java
import java.util.Arrays;

public class sortColors {

    public static void SortColors(int[] nums) {
        if (nums.length <= 1) {
            return;
        }
        int mid = nums.length / 2;
        int[] leftArr = Arrays.copyOfRange(nums, 0, mid);
        int[] rightArr = Arrays.copyOfRange(nums, mid, nums.length);

        SortColors(leftArr);
        SortColors(rightArr);
        merger(nums, leftArr, rightArr);
    }

    public static void merger(int[] nums, int[] leftArr, int[] rightArr) {
        int i = 0, l = 0, r = 0;

        while (l < leftArr.length && r < rightArr.length) {
            if (leftArr[l] < rightArr[r]) {
                nums[i++] = leftArr[l++];
            } else {
                nums[i++] = rightArr[r++];
            }
        }
        while (l < leftArr.length) {
            nums[i++] = leftArr[l++];
        }
        while (r < rightArr.length) {
            nums[i++] = rightArr[r++];
        }
    }

    public static void main(String[] args) {
        /*
         * Given an array nums with n objects colored red, white, or blue, sort them
         * in-place so that objects of the same color are adjacent, with the colors in
         * the order red, white, and blue.
         * 
         * We will use the integers 0, 1, and 2 to represent the color red, white, and
         * blue, respectively.
         * 
         * You must solve this problem without using the library's sort function.
         */
        int[] nums = { 2, 0, 2, 1, 1, 0 };

        SortColors(nums);

        for (int i = 0; i < nums.length; i++) {
            System.out.print(nums[i] + " ");
        }
    }
}